File: generate.c
#include <stdio.h>
#include <getopt.h>

#define VERSION_H                \
"#ifndef VERSION_H\n"            \
"#define VERSION_H\n"            \
"\n"                             \
"#define %s_MAJOR %d\n"          \
"#define %s_MINOR %d\n"          \
"#define %s_PATCH %d\n"          \
"#define %s_BUILD %d\n"          \
"#define %s_META \"%s\"\n"       \
"#define %s \"%d.%d.%d.%d%s\"\n" \
"\n"                             \
"#endif /* VERSION_H */\n"

#define USAGE                                            \
"Usage: %s [-hMmpbaP]\n"                                 \
"\n"                                                     \
"Options:\n"                                             \
"  -h, --help     Print this menu and exit.\n"           \
"  -M, --major    Major number.\n"                       \
"  -m, --minor    Minor number.\n"                       \
"  -p, --patch    Patch number.\n"                       \
"  -b, --build    Build number.\n"                       \
"  -a, --meta     Meta information.\n"                   \
"  -P, --prefix   The macro prefix [default: VERSION]\n" \
"\n"

int main(int argc, char **argv)
{
	struct option long_options[] = {
		{"help",   required_argument, NULL, 'h'},
		{"major",  required_argument, NULL, 'M'},
		{"minor",  required_argument, NULL, 'm'},
		{"patch",  required_argument, NULL, 'p'},
		{"build",  required_argument, NULL, 'b'},
		{"meta",   required_argument, NULL, 'a'},
		{"prefix", required_argument, NULL, 'P'},
	};
	int major, minor, patch, build, c;
	char *meta, *prefix;

	major = minor = patch = build = 0;
	meta = "";
	prefix = "VERSION";

	while ((c = getopt_long(argc, argv, "hM:m:p:b:a:P:", long_options, 0)) != -1) {
		switch (c) {
		case 'h':
			printf(USAGE, argv[0]);
			return 0;
		case 'M':
			major = atoi(optarg);
			break;
		case 'm':
			minor = atoi(optarg);
			break;
		case 'p':
			patch = atoi(optarg);
			break;
		case 'b':
			build = atoi(optarg);
			break;
		case 'a':
			meta = optarg;
			break;
		case 'P':
			prefix = optarg;
			break;
		}
	}

	printf(VERSION_H,
		prefix, major,
		prefix, minor,
		prefix, patch,
		prefix, build,
		prefix, meta,
		prefix, major, minor, patch, build, meta);

	return 0;
}

File: Python_example/generate.py
'''
USAGE:
  generate.py [-o <dir>] <config>

OPTIONS:
  -v, --version                            Print the version
  -h, --help                               Print this help menu
  -o, --output <dir>                       Write output files to this directory
'''
import os
import sys
import json
from textwrap import dedent
from datetime import datetime
from docopt import docopt

class CppGenerator:
    def __init__(self, version):
        self.decorated = version

    def header(self):
        now = datetime.now()
        return dedent('''\
            #ifndef {prefix}VERSION_H
            #define {prefix}VERSION_H

            #define {prefix}VERSION_MAJOR {major}
            #define {prefix}VERSION_MINOR {minor}
            #define {prefix}VERSION_PATCH {patch}
            #define {prefix}VERSION_BUILD {build}
            #define {prefix}VERSION_META "{meta}"
            #define {prefix}VERSION "{major}.{minor}.{patch}.{build}{meta}"
            #define {prefix}COMPILE_DATE "{date}"
            #define {prefix}COMPILE_TIME "{time}"

            const char * hello_world();

            #endif /* {prefix}VERSION_H */
            '''.format(
                prefix = self.decorated['prefix'] if 'prefix' in self.decorated else '',
                major = self.decorated['major'] if 'major' in self.decorated else 0,
                minor = self.decorated['minor'] if 'minor' in self.decorated else 0,
                patch = self.decorated['patch'] if 'patch' in self.decorated else 0,
                build = self.decorated['build'] if 'build' in self.decorated else 0,
                meta = self.decorated['meta'] if 'meta' in self.decorated else '',
                date = now.strftime('%d/%m/%Y'),
                time = now.strftime('%H:%M:%S')
            ))

    def source(self):
        return dedent('''\
            #include "version.h"

            const char * hello_world()
            {
                return "Hello World";
            } 
            ''')

def main():
    argv = docopt(__doc__, version="1.0")

    if argv['<config>'] is not None:
        with open(argv['<config>']) as config_file:
            config = json.load(config_file)
            gen = CppGenerator(config)
            if argv['--output']:
                header_file = os.path.join(argv['--output'], 'version.h')
                source_file = os.path.join(argv['--output'], 'version.c')
                with open(header_file, 'w') as out:
                    out.write(gen.header())
                with open(source_file, 'w') as out:
                    out.write(gen.source())
            else:
                print(gen.header())
                print(gen.source())

if __name__ == "__main__":
    main()
